White label apps branching and workflow strategies (#1201)
codemagic-ci-cd/codemagic-docs: content/knowledge-white-label/white-label-branching-strategies.md

---
description: White label branching and workflow strategies
title: White label branching and workflow strategies
weight: 3
---

There are many different ways you can set up your branching strategy and workflows for white labeling apps. What follows is a recommendation for getting started which you can use as the basis of your own strategy.

## Dev branch and workflow

This "dev" branch is for developing and fixing the core version of the app. The app is built using the default “dev” icons, colors, fonts, bundle id etc. and no white label automation scripts are run to change this app. 

> When new code is committed to this branch it triggers a "dev-release" workflow that runs unit or integration tests. If the tests pass then build the core app and developers can download directly from the Codemagic build page or shared dashboard, via email or Slack notifications, Testflight, Google Play or other distribution channels. 

Reviewer: ```suggestion
When new code is committed to this branch, it triggers a "dev-release" workflow that runs unit or integration tests. If the tests pass, the core app is built and developers can download it directly from the Codemagic build page or shared dashboard, via email or Slack notifications, Testflight, Google Play, or other distribution channels. 
```

> If any of integration or unit tests fail, the developers should review the code, fix and issues and commit new code to trigger a new build.

Reviewer: ```suggestion
If any of integration or unit tests fail, the developers should review the code, fix any issues and commit new code to trigger a new build.
```

{{< /mermaid >}}

## QA branch and workflow

If you are happy with the build from the "dev" branch, you can now proceed to raise a PR request to merge this into the "QA" branch. This triggers a "qa-release" workflows that runs the white label automation scripts to change the icons, images, fonts etc. to something other than the default dev version of the app. The resulting build can be distributed directly to your QA testers. 

{{< mermaid >}}
flowchart TD

%% Colors %%
classDef red fill:#ed2633,stroke:#000,stroke-width:2px,color:#fff

DEV(Dev Branch) ---> PR(Pull Request) ---> BRANCH(QA Branch)
BRANCH ---> TRIGGER(Trigger workflow) 
TRIGGER ---> ANALYSIS(Code analysis)

ANALYSIS ---> UNIT(Unit or E2E tests) 
UNIT ---- PASS(Pass) 
UNIT ---- FAIL(Fail):::red
FAIL --> REVIEW(Review) ---> DEV

PASS --> ASSETS(Get assets & config) ---> SCRIPTS(Run white label scripts) ---> BUILD(Build QA app) 
BUILD ---> DISTRIBUTE(Distribute QA build) ---- TESTFLIGHT(TestFlight) & GOOGLE(Google Play) & FAD(Firebase app distribution) & OTHER(Other)
{{< /mermaid >}}

## Trigger branch and workflow

> The "trigger" branch and workflows serves one purpose; to trigger the builds for each client version. When a pull request is merged into this branch from the "QA" branch it will trigger a "trigger" workflow which is configured to trigger all your client builds using the Codemagic REST API.

Reviewer: ```suggestion
The "trigger" branch and workflow serve one purpose; to trigger the builds for each client version. When a pull request is merged into this branch from the "QA" branch it will trigger a "trigger" workflow which is configured to trigger all your client builds using the Codemagic REST API.
```
